src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -1017,6 +1021,16 @@ public void RemoveOutputPath(string outputPath)
             _workspace.RemoveProjectOutputPath(Id, outputPath);
         }
 
+        public void ReorderSourceFiles(ImmutableArray<string> filePaths)
+        {
+            if (filePaths.IsEmpty)
+            {
+                throw new ArgumentOutOfRangeException(ServicesVSResources.The_specified_file_paths_is_empty, nameof(filePaths));
+            }
+
+            _sourceFiles.ReorderFiles(filePaths);
```

Comment:
So this is a bit strange; it means somebody could start a batch, reorder file paths _that haven't been added_, then add them, and then complete the batch. Seems validation should be happening up front.

Reply:
In the batch Test I wrote, it describes how this interaction works.

```csharp
                project.StartBatch();
                // We try to reorder here with an invalid source file. The next call to ReorderSourceFiles will override this one in a batch.
                project.ReorderSourceFiles(new[] { @"C:\invalid source file" });

                // This one will override the previous call to ReorderSourceFiles in a batch
                // Also, we can call ReorderSourceFiles before we have added all the source files that the ordering will include because we are in a batch.
                project.ReorderSourceFiles(new[] { sourceFileFullPath1, sourceFileFullPath2, sourceFileFullPath3, sourceFileFullPath4, sourceFileFullPath5 });

                project.AddSourceFile(sourceFileFullPath1);
                project.AddSourceFile(sourceFileFullPath2);
                project.AddSourceFile(sourceFileFullPath3);
                project.AddSourceFile(sourceFileFullPath4);
                project.AddSourceFile(sourceFileFullPath5);

                project.RemoveSourceFile(sourceFileFullPath2);
                project.AddSourceFile(sourceFileFullPath2);

                project.RemoveSourceFile(sourceFileFullPath4);
                project.AddSourceFile(sourceFileFullPath4);

                project.EndBatch();
```

Reply:
Do we not want this to behave this way? I can only reorder source files until EndBatch has been called that actually makes changes to ProjectState because I rely on ProjectState for reordering files. I might be able to do some special validation for batching if this is not the behavior we want.

Reply:
Upon looking into validation for this, it might get a little complicated. We would need to change how batching works a bit.

Today if you do this in a batch:
```csharp
project.StartBatch();

project.AddSourceFile(file1);
project.RemoveSourceFile(file1);
project.AddSourceFile(file1);

project.EndBatch();
```
What is the behavior in this case?

Reply:
The first RemoveSourceFile should remove the added-to-a-batch-but-not-really-added file, and then the second AddSourceFile should complete. Basically, the model was if you removed the StartBatch/EndBatch, the code would behave the same -- any problems are validated up front and EndBatch doesn't throw.